test/test_testing.py

```diff
@@ -123,6 +125,119 @@ def test_update_relation_exposes_new_data(self):
         self.assertEqual(viewer.changes, [{'initial': 'data'},
                                           {'initial': 'data', 'new': 'value'}])
 
+    def test_update_relation_no_local_unit_change_event(self):
+        # language=YAML
+        harness = Harness(CharmBase, meta='''
+            name: my-charm
+            requires:
+              db:
+                interface: pgsql
+            ''')
+        harness.begin()
+        helper = DBRelationChangedHelper(harness.charm, "helper")
+        rel_id = harness.add_relation('db', 'postgresql')
+        with self.assertRaises(RuntimeError):
+            harness.update_relation_data(rel_id, 'my-charm/0', {'new': 'value'})
+        rel = harness.charm.model.get_relation('db')
+        # Make sure our unit data has not actually changed after an exception got raised.
+        with self.assertRaises(KeyError):
+            self.assertEqual(rel.data[harness.charm.model.unit]['new'], 'value')
+
+        rel.data[harness.charm.model.unit]['new'] = 'value'
+        # Our unit data bag got updated.
+        self.assertEqual(rel.data[harness.charm.model.unit]['new'], 'value')
+        # But there were no changed events registered by our unit.
+        self.assertEqual([], helper.changes)
+
+    def test_update_peer_relation_no_local_unit_change_event(self):
+        # language=YAML
+        harness = Harness(CharmBase, meta='''
+            name: postgresql
+            peers:
+              db:
+                interface: pgsql
+            ''')
+        harness.begin()
+        helper = DBRelationChangedHelper(harness.charm, "helper")
+        rel_id = harness.add_relation('db', 'postgresql')
+
+        with self.assertRaises(RuntimeError):
+            harness.update_relation_data(rel_id, 'postgresql/0', {'k1': 'v1'})
+        rel = harness.charm.model.get_relation('db')
+        # Make sure our unit data has not actually changed after an exception got raised.
+        with self.assertRaises(KeyError):
+            self.assertEqual(rel.data[harness.charm.model.unit]['k1'], 'v1')
+
+        rel.data[harness.charm.model.unit]['k1'] = 'v1'
+        # Our unit data bag got updated.
+        self.assertEqual(rel.data[harness.charm.model.unit]['k1'], 'v1')
+        # But there were no changed events registered by our unit.
+        self.assertEqual([], helper.changes)
+
+        # Same for when our unit is a leader.
+        harness.set_leader(is_leader=True)
+        with self.assertRaises(RuntimeError):
+            harness.update_relation_data(rel_id, 'postgresql/0', {'k2': 'v2'})
+        self.assertEqual(rel.data[harness.charm.model.unit]['k1'], 'v1')
```

Comment:
This test is actually looking at the wrong keys.

Maybe instead of KeyError we just look at the databag as a whole? We might need dict() as we have a custom type, but I think:

self.assertEqual({'k1':'v1'}, dict(rel.data[harness.charm.model.unit]))

Reply:
Yes, you are right, thanks for catching.

Reply:
https://github.com/canonical/operator/pull/212/commits/2e440c669c1d1f89186418f2f194494bae9389f9